src/modules/ssh_apt/mappers/index.ts

```diff
@@ -0,0 +1,130 @@
+import { SshAptModule } from '..';
+import { Context, Crud, MapperBase } from '../../interfaces';
+import { Package } from '../entity';
+
+export class PackageMapper extends MapperBase<Package> {
+  module: SshAptModule;
+  entity = Package;
+
+  equals = (a: Package, b: Package) =>
+    Object.is(a.server, b.server) &&
+    Object.is(a.package, b.package) &&
+    Object.is(a.version, b.version) &&
+    Object.is(a.architecture, b.architecture) &&
+    Object.is(a.description, b.description) &&
+    Object.is(a.installed, b.installed) &&
+    Object.is(a.upgradable, b.upgradable);
+
+  cloud = new Crud<Package>({
+    create: async (es: Package[], ctx: Context) => {
+      // Users cannot create packages, only install packages provided by their sources, so simply
+      // delete these records
+      const out = await this.module.package.db.delete(es, ctx);
+      if (Array.isArray(out) || !out) return out;
+      return [out];
+    },
+    read: async (ctx: Context, id?: string) => {
+      if (id) {
+        // There's no good way to read a single value, so we're going to read *all the things*
+        // and then filter out the desired record
+        const packages: Package[] = (await this.module.package.cloud.read(ctx)) ?? [];
+        const { server, package: deb, version, architecture } = this.idFields(id);
+        return packages.find(
+          (p: Package) =>
+            p.server === server &&
+            p.package === deb &&
+            p.version === version &&
+            p.architecture === architecture,
+        );
+      }
+      const servers = await ctx.getServerNames();
+      const clientsAndServers = await Promise.all(
+        servers.map(async (s: string) => ({
+          server: s,
+          client: await ctx.getSshClient(s),
+        })),
+      );
+      const out: Package[] = [];
+      await Promise.all(
+        clientsAndServers.map(async clientAndServer => {
+          const { client, server } = clientAndServer;
+          const allPackages = (await client.exec('sudo apt list --all-versions --verbose 2>/dev/null | cat'))
+            .split('\n')
+            .slice(1)
+            .filter((line: string) => !!line)
+            .map((line: string) => line.trim());
+          const installedPackages = (await client.exec('sudo apt list --installed 2>/dev/null | cat'))
+            .split('\n')
+            .slice(1)
+            .filter((line: string) => !!line);
+          const upgradablePackages = (await client.exec('sudo apt list --upgradable 2>/dev/null | cat'))
+            .split('\n')
+            .slice(1)
+            .filter((line: string) => !!line);
```

Comment:
Wondering if there's an npm package that can do these tasks over SSH?

Reply:
I looked. I didn't see one, but I don't think this code is that bad?